stetho/src/main/java/com/facebook/stetho/inspector/elements/Document.java

```diff
@@ -0,0 +1,483 @@
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Queue;
+
+public class Document {
+  private DOMProvider.Factory mFactory;
+
+  @Nullable
+  private DOMProvider mDOMProvider;
+  private ShadowDOM mShadowDOM;
+  private ObjectIdMapper mObjectIdMapper;
+  private UpdateListener mUpdateListener;
+
+  private AttributeListAccumulator mCachedAttributeAccumulator;
+  private ChildEventingList mCachedChildEventingList;
+  private ArrayListAccumulator<Object> mCachedChildrenAccumulator;
+
+  private int mReferenceCounter;
+
+  public Document(DOMProvider.Factory factory) {
+    mFactory = factory;
+    mObjectIdMapper = new DOMObjectIdMapper();
+    mReferenceCounter = 0;
+  }
+
+  public synchronized void AddRef() {
+    if (mReferenceCounter++ == 0) {
+      mDOMProvider = mFactory.create();
+    }
+  }
+
+  public synchronized void Release() {
+    if (mReferenceCounter > 0) {
+      if (--mReferenceCounter == 0) {
+        cleanUp();
+      }
+    }
+  }
+
+  public void setUpdateListener(UpdateListener updateListener) {
+    mUpdateListener = updateListener;
+  }
+
+  public ShadowDOM getShadowDOM() {
+    return mShadowDOM;
+  }
+
+  public DOMProvider getDOMProvider() {
+    return mDOMProvider;
+  }
+
+  public ObjectIdMapper getObjectIdMapper() {
+    return mObjectIdMapper;
+  }
+
+  public Object getRootElement() {
+    return mDOMProvider.postAndWait(new UncheckedCallable<Object>() {
+      @Override
+      public Object call() {
+        Object rootElement = mDOMProvider.getRootElement();
+        if (rootElement == null) {
+          // null for rootElement is not allowed. We could support it, but our current
+          // implementation won't ever run into this, so let's punt on it for now.
+          throw new IllegalStateException();
+        }
+
+        if (mShadowDOM == null) {
+          mShadowDOM = new ShadowDOM(mDOMProvider.getRootElement());
+          createShadowDOMUpdate().commit();
+          mDOMProvider.setListener(new ProviderListener());
+        } else if (rootElement != mShadowDOM.getRootElement()) {
+          // We don't support changing the root element. This is handled differently by the
+          // protocol than updates to an existing DOM, and we don't have any case in our
+          // current implementation that causes this to happen, so let's punt on it for now.
+          throw new IllegalStateException();
+        }
+
+        return rootElement;
+      }
+    });
+  }
+
+  public Object getElement(int id) {
+    return mObjectIdMapper.getObjectForId(id);
+  }
+
+  public String[] getElementAttributes(Object element) {
+    final Document.AttributeListAccumulator accumulator = getCachedAttributeListAccumulator();
+    NodeDescriptor descriptor = mDOMProvider.getNodeDescriptor(element);
+    descriptor.getAttributes(element, accumulator);
+
+    String[] result = accumulator.toArray(new String[accumulator.size()]);
```

Comment:
we should not be returning an array. let the caller send in an AttributeAccumulator.